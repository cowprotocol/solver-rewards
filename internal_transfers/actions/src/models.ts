// Event data declared here consists of only the relevant fields
// for our purposes at this time.
// For example, this program only requires the owner of the trade so
// that transfers can be classified AMM_{IN/OUT} or USER_{IN/OUT}.

export interface PipelineSecrets {
  nodeUrl: string;
  dbUrl: string;
  simulatorKey: string;
}

export interface TokenImbalance {
  token: string;
  amount: bigint;
}

export interface Trace {
  callType: string;
  from: string;
  to: string;
  // This is a hex representation of a bigint
  value: string;
}

export interface EventLog {
  // Block-wise index of Log
  readonly index: number;
  // The contract address emitting the event
  readonly address: string;
  // The indexed topics from the event log
  readonly topics: string[];
  // The additional (non-indexed event data)
  readonly data: string;
}

export interface TradeEvent {
  owner: string;
}

export function isTradeEvent(event: any): event is TradeEvent {
  const castEvent = event as TradeEvent;
  return castEvent.owner !== undefined;
}
export interface TransferEvent {
  token: string;
  to: string;
  from: string;
  amount: bigint;
}
export interface EventMeta {
  blockNumber: number;
  txHash: string;
}

export interface SettlementEvent {
  solver: string;
  logIndex: number;
}

export function isSettlementEvent(event: any): event is SettlementEvent {
  const castEvent = event as SettlementEvent;
  return castEvent.solver !== undefined && castEvent.logIndex !== undefined;
}

export interface WinningSettlementData {
  // Block at which settlement was simulated.
  simulationBlock: number;
  // Solver who submitted the (winning) solution
  solver: string;
  // Settlement call data: after internalized interactions have been removed.
  // equivalent to what actually appears on-chain.
  reducedCallData: string;
  // Full Call Data provided by the solver
  // null indicates that settlement was not pruned.
  fullCallData?: string;
}
