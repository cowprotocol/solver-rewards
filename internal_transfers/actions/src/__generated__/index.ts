/**
 * !!! This file is autogenerated do not edit by hand !!!
 *
 * Generated by: @databases/pg-schema-print-types
 * Checksum: /1vc9MIUeiqemAs/y83Z4Au8FsEWZiaqCnNE8ZI22e7QJgWX5wevKh8UrupohKHYF5vbMzfegV1/agHqfr8FNA==
 */

/* eslint-disable */
// tslint:disable

import FlywaySchemaHistory, {
  FlywaySchemaHistory_InsertParameters,
} from "./flyway_schema_history";
import InternalizedImbalances, {
  InternalizedImbalances_InsertParameters,
} from "./internalized_imbalances";
import SettlementSimulations, {
  SettlementSimulations_InsertParameters,
} from "./settlement_simulations";
import Settlements, { Settlements_InsertParameters } from "./settlements";
import TxReceipts, { TxReceipts_InsertParameters } from "./tx_receipts";

interface DatabaseSchema {
  flyway_schema_history: {
    record: FlywaySchemaHistory;
    insert: FlywaySchemaHistory_InsertParameters;
  };
  internalized_imbalances: {
    record: InternalizedImbalances;
    insert: InternalizedImbalances_InsertParameters;
  };
  settlement_simulations: {
    record: SettlementSimulations;
    insert: SettlementSimulations_InsertParameters;
  };
  settlements: { record: Settlements; insert: Settlements_InsertParameters };
  tx_receipts: { record: TxReceipts; insert: TxReceipts_InsertParameters };
}
export default DatabaseSchema;

/**
 * JSON serialize values (v) if the table name (t) and column name (c)
 * is a JSON or JSONB column.
 * This is necessary if you want to store values that are not plain objects
 * in a JSON or JSONB column.
 */
function serializeValue(t: string, c: string, v: unknown): unknown {
  if (
    (t === "settlement_simulations" &&
      (c === "complete" || c === "reduced" || c === "winning_settlement")) ||
    (t === "tx_receipts" && c === "data")
  ) {
    return JSON.stringify(v);
  }
  return v;
}
export { serializeValue };

export type {
  FlywaySchemaHistory,
  FlywaySchemaHistory_InsertParameters,
  InternalizedImbalances,
  InternalizedImbalances_InsertParameters,
  SettlementSimulations,
  SettlementSimulations_InsertParameters,
  Settlements,
  Settlements_InsertParameters,
  TxReceipts,
  TxReceipts_InsertParameters,
};
