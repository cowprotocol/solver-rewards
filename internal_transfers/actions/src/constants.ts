export const SETTLEMENT_CONTRACT_ADDRESS =
  "0x9008D19f58AAbD9eD0D60971565AA8510560ab41";
