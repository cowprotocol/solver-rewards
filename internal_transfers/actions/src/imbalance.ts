import { TokenImbalance, TransferEvent } from "./models";

export type ImbalanceMap = Map<string, bigint>;

/**
 * Aggregation operator transforming a list of `TransferEvent` into an
 * aggregation of token imbalances as an `ImbalanceMap`.
 * Imbalances are defined in reference to `focalAccount`
 * @param transfers - a list of token transfers
 * @param focalAccount - account for which the "imbalance" is referring to.
 */
export function aggregateTransfers(
  transfers: TransferEvent[],
  focalAccount: string
): ImbalanceMap {
  const accumulator: ImbalanceMap = new Map<string, bigint>();
  transfers.map((transfer) => {
    let { to, from, amount, token } = transfer;
    token = token.toLowerCase();
    const currVal = accumulator.get(token) ?? 0n;
    if (to.toLowerCase() == focalAccount.toLowerCase()) {
      // Incoming Transfer
      accumulator.set(token, currVal + amount);
    } else if (from.toLowerCase() == focalAccount.toLowerCase()) {
      // Outgoing Transfer
      accumulator.set(token, currVal - amount);
    } else {
      // Irrelevant transfer.
    }
  });
  return accumulator;
}

/**
 * Difference operator on two token imbalance mappings.
 * More generally, this computes the difference of two mappings of type { string => bigint }.
 * Example: ({ "a": 1, "b": 2 }, { "b": 3, "c": 4 }) --> { "a": 1, "b": -1, "c": -4 }
 * @param mapA positive term - being subtracted from
 * @param mapB negative term - being subtracted
 * @return mapA - mapB as a list of TokenImbalance
 */
export function imbalanceMapDiff(
  mapA: ImbalanceMap,
  mapB: ImbalanceMap
): TokenImbalance[] {
  const keySet = new Set([...mapA.keys(), ...mapB.keys()]);
  let diffMap: ImbalanceMap = new Map<string, bigint>();
  for (const token of keySet) {
    const difference = (mapA.get(token) ?? 0n) - (mapB.get(token) ?? 0n);
    if (difference !== 0n) {
      // No point in recording zeros!
      diffMap.set(token, difference);
    }
  }
  return Array.from(diffMap).map(([token, amount]) => ({
    token,
    amount,
  }));
}
