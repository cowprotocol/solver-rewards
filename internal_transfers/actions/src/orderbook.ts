import { WinningSettlementData } from "./models";
import axios from "axios";

// https://api.cow.fi/docs/#/
const PROD_API_URL =
  "https://api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash";
const BARN_API_URL =
  "https://barn.api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash";

async function getWinningSettlementFromOrderbookAPI(
  apiUrl: string
): Promise<WinningSettlementData | undefined> {
  try {
    const response = await axios.get(apiUrl);

    // The Orderbook, stores all solution submissions sorted by the objective criteria.
    // The winning solution is the last entry of the `solutions` array.
    const solutionArray = response.data.solutions;
    const winningSolution = solutionArray[solutionArray.length - 1];
    // TODO - Note that old records have solver = ZeroAddress!
    //  We need to rectify this (use txReceipt or whatever Tenderly web-action provides)
    const solver = winningSolution.solverAddress;

    return {
      simulationBlock: response.data.competitionSimulationBlock,
      solver,
      reducedCallData: winningSolution.callData,
      fullCallData: winningSolution.uninternalizedCallData,
    };
  } catch (exception) {
    if (axios.isAxiosError(exception)) {
      if (exception.response?.status != 404) {
        throw new Error(
          `ERROR response status from ${apiUrl}: ${exception.response}`
        );
      }
      // Not Found (404) is an expected response from the API.
      // Could be for any of the following reasons:
      // 1. !isValidTxHash (regex:  /^0x([A-Fa-f0-9]{64})$/ )
      // 2. !isActualTxHash: web3.eth.getTransactionReceipt(txHash) -- Requires EVM Node.
      // 3. !isSettlementTransaction(txReceipt)
      // 4. !correctDB
      // TODO - might be nice to pre-validate input, but would affect performance of workflow
      //  its cheaper and faster to query orderbook API for the data we want without validation.
      return undefined;
    } else {
      throw new Error(
        `ERROR unexpected response received from ${apiUrl}: ${exception}`
      );
    }
  }
}

/** Fetches solver_competition for `tx_hash` from orderbook API via:
 * api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{tx_hash}
 * Parses only the relevant fields from the results and returns this info.
 *
 * @param {string}  txHash - transaction hash of a settlement event.
 * @returns {WinningSettlementData}
 */
export async function getSettlementCompetitionData(
  txHash: string
): Promise<WinningSettlementData | undefined> {
  // Try fetching from Production first
  let result = await getWinningSettlementFromOrderbookAPI(
    `${PROD_API_URL}/${txHash}`
  );
  if (result !== undefined) return result;

  // If it was not found/undefined try fetching from Barn.
  return await getWinningSettlementFromOrderbookAPI(
    `${BARN_API_URL}/${txHash}`
  );
}
