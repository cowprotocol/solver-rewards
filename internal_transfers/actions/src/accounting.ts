import { SETTLEMENT_CONTRACT_ADDRESS } from "./constants";
import { getSettlementCompetitionData } from "./orderbook";
import { SimulationData, TransactionSimulator } from "./simulate/interface";
import { partitionEventLogs } from "./parse";
import { TokenImbalance, WinningSettlementData, EventLog } from "./models";
import {
  aggregateTransfers,
  ImbalanceMap,
  imbalanceMapDiff,
} from "./imbalance";

const ETH_TOKEN = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee";

// This represents the only required fields for InternalImbalance accounting
// An object of this shape can be easily be constructed from any of
// - ethers - `TransactionReceipt`,
// - tenderly - `TransactionEvent`
export interface MinimalTxData {
  readonly blockNumber: number;
  readonly from: string;
  readonly hash: string;
  readonly logs: EventLog[];
}

export function constructImbalanceMap(
  simulation: SimulationData,
  focalContract: string
): ImbalanceMap {
  focalContract = focalContract.toLowerCase();
  const { transfers } = partitionEventLogs(simulation.logs);

  const simulationImbalance = aggregateTransfers(transfers, focalContract);
  const ethImbalance = simulation.ethDelta.get(focalContract) ?? 0n;
  if (ethImbalance !== 0n) {
    simulationImbalance.set(ETH_TOKEN, ethImbalance);
  }

  return simulationImbalance;
}

export interface SettlementSimulationData {
  txHash: string;
  winningSettlement: WinningSettlementData;
  full: SimulationData;
  reduced: SimulationData;
}

export async function simulateSolverSolution(
  transaction: MinimalTxData,
  simulator: TransactionSimulator
): Promise<SettlementSimulationData | null> {
  const solverAddress = transaction.from.toLowerCase();
  const competition = await getSettlementCompetitionData(transaction.hash);
  if (competition === undefined) {
    throw Error(`No competition found for ${transaction.hash}`);
  }

  if (competition.fullCallData === undefined) {
    // Settlement was not even partially internalized.
    // https://api.cow.fi/docs/#/default/get_api_v1_solver_competition_by_tx_hash__tx_hash_
    // No need to simulate!
    console.log(`batch ${transaction.hash} was not internalized.`);
    return null;
  }
  // Try all blocks between claimed simulation block and mined block.
  const numAttempts = transaction.blockNumber - competition.simulationBlock + 1;
  try {
    const { full, reduced } = await simulateBoth(
      simulator,
      {
        full: competition.fullCallData,
        reduced: competition.reducedCallData,
        common: {
          contractAddress: SETTLEMENT_CONTRACT_ADDRESS,
          sender: solverAddress,
          value: "0",
        },
        startBlock: competition.simulationBlock,
      },
      numAttempts
    );
    return {
      txHash: transaction.hash,
      winningSettlement: competition,
      full,
      reduced,
    };
  } catch (error: any) {
    console.error(error.message);
    // Sometimes (rarely) we can't simulate both components of the solver competition data.
    // When this happens, it is assumed that there were no internalized transfers
    // and write a kind of placeholder/trivial record as follows:
    const failedSimulation = {
      simulationID: `failed all ${numAttempts} simulation attempts`,
      blockNumber: -1, // easily identifiable "trivial simulation record"
      gasUsed: 0, // 0 will not affect aggregate sums on gas consumption (i.e. analytics)
      logs: [], // implies no token transfers.
      ethDelta: new Map(), // implies no eth balance diff.
    };
    return {
      txHash: transaction.hash,
      winningSettlement: competition,
      full: failedSimulation,
      reduced: failedSimulation,
    };
  }
}

interface commonSimulationParams {
  contractAddress: string;
  sender: string;
  value: string;
}

interface SettlementSimulationParams {
  full: string;
  reduced: string;
  common: commonSimulationParams;
  startBlock: number;
}

interface SimulationPair {
  full: SimulationData;
  reduced: SimulationData;
}

async function simulateBoth(
  simulator: TransactionSimulator,
  params: SettlementSimulationParams,
  numAttempts: number = 3
): Promise<SimulationPair> {
  let attemptNumber = 0;
  while (attemptNumber < numAttempts) {
    try {
      return {
        full: await simulator.simulate({
          ...params.common,
          blockNumber: params.startBlock + attemptNumber,
          callData: params.full,
        }),
        reduced: await simulator.simulate({
          ...params.common,
          blockNumber: params.startBlock + attemptNumber,
          callData: params.reduced,
        }),
      };
    } catch (error) {
      attemptNumber += 1;
      console.warn(`Failed simulation attempt ${attemptNumber}`);
    }
  }
  throw new Error(
    `failed simulations on ${numAttempts} blocks beginning from ${params.startBlock}`
  );
}

export function getInternalizedImbalance(
  simulationData: SettlementSimulationData
): TokenImbalance[] {
  const fullSimulationImbalance = constructImbalanceMap(
    simulationData.full,
    SETTLEMENT_CONTRACT_ADDRESS
  );
  const reducedSimulationImbalance = constructImbalanceMap(
    simulationData.reduced,
    SETTLEMENT_CONTRACT_ADDRESS
  );

  return imbalanceMapDiff(fullSimulationImbalance, reducedSimulationImbalance);
}
