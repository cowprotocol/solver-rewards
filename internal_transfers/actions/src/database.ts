// Code Reference: https://www.atdatabases.org/docs/pg-guide-typescript

import createConnectionPool, { Queryable, sql } from "@databases/pg";
import tables from "@databases/pg-typed";
import ConnectionPool from "@databases/pg/lib/types/Queryable";
import DatabaseSchema from "./__generated__";
import { EventMeta, SettlementEvent, TokenImbalance } from "./models";
import { MinimalTxData, SettlementSimulationData } from "./accounting";
import { SimulationData } from "./simulate/interface";

export { sql };
const {
  settlements,
  internalized_imbalances,
  settlement_simulations,
  tx_receipts,
} = tables<DatabaseSchema>({
  databaseSchema: require("./__generated__/schema.json"),
});

function pgHash(txHash: string) {
  return txHash.replace("0x", "\\x");
}

function getDB(dbURL: string): ConnectionPool {
  return createConnectionPool({
    connectionString: dbURL,
    bigIntMode: "bigint",
  });
}

export async function insertTxReceipt(
  db: ConnectionPool,
  receipt: MinimalTxData
) {
  await tx_receipts(db).insertOrIgnore({
    hash: hexToBytea(receipt.hash),
    block_number: receipt.blockNumber,
    data: receipt,
  });
}

export async function getUnprocessedReceipts(
  db: Queryable,
  blockFrom: number
): Promise<MinimalTxData[]> {
  const query = sql`SELECT data from tx_receipts where processed = false and block_number < ${blockFrom};`;
  const unstructuredResults = await db.query(query);
  return unstructuredResults.map((value) => {
    return value.data as MinimalTxData;
  });
}

export async function markReceiptProcessed(db: Queryable, hash: string) {
  const updateQuery = sql`UPDATE tx_receipts SET processed = true where hash = ${pgHash(
    hash
  )};`;
  await db.query(updateQuery);
}

export async function recordExists(
  db: Queryable,
  txHash: string
): Promise<boolean> {
  const query = sql`SELECT count(*) from settlements where tx_hash = ${pgHash(
    txHash
  )};`;
  const { count: numRecords } = (await db.query(query))[0];
  return numRecords > 0;
}

async function insertSettlementEvent(
  db: ConnectionPool,
  eventMeta: EventMeta,
  settlementEvent: SettlementEvent
) {
  const { solver, logIndex } = settlementEvent;
  const { txHash, blockNumber } = eventMeta;
  await settlements(db).insert({
    tx_hash: hexToBytea(txHash),
    solver: hexToBytea(solver),
    block_number: blockNumber,
    log_index: logIndex,
  });
}

async function insertTokenImbalances(
  db: ConnectionPool,
  txHash: string,
  imbalances: TokenImbalance[]
) {
  await internalized_imbalances(db).bulkInsert({
    columnsToInsert: ["token", "tx_hash", "amount"],
    records: imbalances.map((imbalance) => {
      return {
        token: hexToBytea(imbalance.token),
        tx_hash: hexToBytea(txHash),
        amount: imbalance.amount.toString(),
      };
    }),
  });
}

export function bigIntMapToJSON(originalMap: Map<string, bigint>): object {
  return Object.fromEntries(
    new Map(Array.from(originalMap, ([k, v]) => [k, v.toString()]))
  );
}
export function jsonFromSettlementData(datum: SimulationData): object {
  return {
    logs: datum.logs,
    blockNumber: datum.blockNumber,
    ethDelta: bigIntMapToJSON(datum.ethDelta),
    simId: datum.simulationID,
  };
}

async function insertSettlementSimulations(
  db: ConnectionPool,
  datum: SettlementSimulationData
) {
  await settlement_simulations(db).insert({
    tx_hash: hexToBytea(datum.txHash),
    complete: jsonFromSettlementData(datum.full),
    reduced: jsonFromSettlementData(datum.reduced),
    winning_settlement: datum.winningSettlement,
  });
}

export interface SlippagePipelineResults {
  settlementSimulations: SettlementSimulationData;
  imbalances: TokenImbalance[];
  eventMeta: EventMeta;
  settlementEvent: SettlementEvent;
}

export async function insertSettlementAndMarkProcessed(
  db: ConnectionPool,
  eventMeta: EventMeta,
  settlementEvent: SettlementEvent
) {
  await db.tx(async (db) => {
    await insertSettlementEvent(db, eventMeta, settlementEvent);
    await markReceiptProcessed(db, eventMeta.txHash);
  });
}

export async function insertPipelineResults(
  db: ConnectionPool,
  pipelineResults: SlippagePipelineResults
) {
  const { eventMeta, settlementEvent, imbalances, settlementSimulations } =
    pipelineResults;
  await db.tx(async (db) => {
    await insertTokenImbalances(db, eventMeta.txHash, imbalances);
    await insertSettlementSimulations(db, settlementSimulations);
    await insertSettlementAndMarkProcessed(db, eventMeta, settlementEvent);
  });
  console.log(`wrote ${imbalances.length} imbalances for ${eventMeta.txHash}`);
}

function hexToBytea(hexString: string): Buffer {
  return Buffer.from(hexString.slice(2), "hex");
}

export {
  insertSettlementEvent,
  insertTokenImbalances,
  getDB,
  hexToBytea,
  insertSettlementSimulations,
};
