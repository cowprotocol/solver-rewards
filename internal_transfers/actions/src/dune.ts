import { QueryParameter, DuneClient } from "@cowprotocol/ts-dune-client";
import { MinimalTxData } from "./accounting";

const TX_DATA_QUERY_ID = 2492342;

export async function getSampleSet(
  dune: DuneClient,
  // YYYY-MM-DD
  dateFrom: string,
  dateTo: string
): Promise<MinimalTxData[]> {
  const historicalTransactionData = await dune
    .refresh(TX_DATA_QUERY_ID, [
      QueryParameter.date("DateFrom", `${dateFrom} 00:00:00`),
      QueryParameter.date("DateTo", `${dateTo} 00:00:00`),
    ])
    .then((executionResult) => executionResult.result);
  if (!historicalTransactionData) {
    throw new Error("Failure");
  }

  return historicalTransactionData.rows.map((row: Record<string, any>) => {
    const { blockNumber, from, hash, logs } = row;
    if (!isEventLogArray(logs)) {
      throw Error(`Invalid Row ${JSON.stringify(row)}`);
    }
    return {
      blockNumber: parseInt(blockNumber),
      from,
      hash,
      logs: logs.map((log) => JSON.parse(log)),
    };
  });
}

function isEventLogArray(value: any): value is Array<string> {
  return value instanceof Array;
}
