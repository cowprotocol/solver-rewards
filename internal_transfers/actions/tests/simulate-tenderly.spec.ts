import {
  isTenderlySimulationResponse,
  parseTenderlySimulation,
  TenderlySimulator,
} from "../src/simulate/tenderly";
import { SETTLEMENT_CONTRACT_ADDRESS } from "../src/constants";
import { config as envConfig } from "dotenv";

const oneAddress = "0x0000000000000000000000000000000000000001";

const invalidSimulator = new TenderlySimulator(
  "INVALID_USER",
  "INVALID_PROJECT",
  "INVALID_KEY"
);
envConfig();
// Note that simulator will also be invalid if the env vars are not set.
const simulator = new TenderlySimulator(
  process.env["TENDERLY_USER"] || "TENDERLY_USER",
  process.env["TENDERLY_PROJECT"] || "TENDERLY_PROJECT",
  process.env["TENDERLY_ACCESS_KEY"] || "TENDERLY_ACCESS_KEY"
);
describe("Tenderly Simulator", () => {
  test.skip("simulate() returns expected output on successful simulation", async () => {
    const simulation = await simulator.simulate({
      callData:
        "0x13d79a0b000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000036000000000000000000000000000000000000000000000000000000000000000020000000000000000000000003432b6a60d23ca0dfca7761b7ab56459d9c964d00000000000000000000000006b175474e89094c44da98b954eedeac495271d0f000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000009e1a51c6b00000000000000000000000000000000000000000000000000000002540be4000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e2b424053b9ebfcedf89ecb8bf2972974e98700c00000000000000000000000000000000000000000000006c0b439bc588511000000000000000000000000000000000000000000000000019629e5fa65ea1706b0000000000000000000000000000000000000000000000000000000063af639ec86d3a0def4d16bd04317645da9ae1d6871726d8adf83a0695447f8ee5c63d12000000000000000000000000000000000000000000000000604fbfc634eef000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006c0b439bc58851100000000000000000000000000000000000000000000000000000000000000001600000000000000000000000000000000000000000000000000000000000000041d6f2b85ab927ca9078c1081587fe4ea7ec72a9e1ef3ed826649e8fa063bafb45643196adc29f3c13403865f063131bb2a5a293483423c810ea56dc764e25e2971c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000260000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000020000000000000000000000000d9e1ce17f2641f24ae83637ab66a2cca9c378b9f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000001048803dbee000000000000000000000000000000000000000000000018ba53f5c7f06ef117000000000000000000000000000000000000000000000000166c0a97dd56bfd500000000000000000000000000000000000000000000000000000000000000a00000000000000000000000009008d19f58aabd9ed0d60971565aa8510560ab41ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc20000000000000000000000003432b6a60d23ca0dfca7761b7ab56459d9c964d0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      sender: "0xa21740833858985e4d801533a808786d3647fb83",
      value: "0",
      contractAddress: SETTLEMENT_CONTRACT_ADDRESS,
      blockNumber: 16300366,
    });
    expect(simulation).toMatchSnapshot();
  });
  test.skip("simulate() returns expected output on failed simulation", async () => {
    await expect(
      simulator.simulate({
        callData: "0x",
        sender: oneAddress,
        value: "0",
        contractAddress: SETTLEMENT_CONTRACT_ADDRESS,
        blockNumber: 16300366,
      })
    ).rejects.toThrow();
  });
  test("simulate() returns expected output with invalid credentials", async () => {
    await expect(
      invalidSimulator.simulate({
        callData: "0x",
        contractAddress: "0x",
        sender: "0x",
        value: "0",
      })
    ).rejects.toThrow("Request failed with status code 403");
  });
  test("simulate() throws bad request error", async () => {
    const badRequestSimulator = new TenderlySimulator("", "", "");
    await expect(
      badRequestSimulator.simulate({
        callData: "0x",
        contractAddress: "0x",
        sender: "0x",
        value: "0",
      })
    ).rejects.toThrow("Request failed with status code 401");
  });

  test("parseTenderlySimulation() operates as expected", () => {
    const mockSimulation = {
      simulation: {
        id: "id-value",
        gas_used: 1,
      },
      // Test Data excludes irrelevant fields.
      transaction: {
        hash: "0xHASH",
        transaction_info: {
          balance_diff: [],
          block_number: 5,
          logs: [
            {
              raw: {
                address: "0xADDRESS",
                data: "0xDATA",
                topics: ["0xTOPIC_1"],
              },
            },
          ],
        },
      },
    };
    expect(parseTenderlySimulation(mockSimulation)).toEqual({
      blockNumber: 5,
      gasUsed: 1,
      ethDelta: new Map(),
      logs: [
        {
          index: 0,
          address: "0xADDRESS",
          data: "0xDATA",
          topics: ["0xTOPIC_1"],
        },
      ],
      simulationID: "tenderly-id-value",
    });
  });
  test("isTenderlySimulationResponse() returns false", () => {
    expect(
      isTenderlySimulationResponse({
        transaction: {
          hash: "",
          transaction_info: {
            block_number: 0,
            // TODO - No Logs... probably shouldn't fail here!
            logs: null,
          },
        },
      })
    ).toBe(false);
  });
});
