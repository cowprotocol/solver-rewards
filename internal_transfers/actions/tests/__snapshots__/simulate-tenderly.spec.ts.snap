// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tenderly Simulator parseSimulation() operates as expected 1`] = `
{
  "blockNumber": 16300366,
  "logs": [
    {
      "address": "0x9008d19f58aabd9ed0d60971565aa8510560ab41",
      "data": "0x0000000000000000000000006b175474e89094c44da98b954eedeac495271d0f0000000000000000000000003432b6a60d23ca0dfca7761b7ab56459d9c964d000000000000000000000000000000000000000000000006c6b935b8bbd400000000000000000000000000000000000000000000000000019753399721b8078ee000000000000000000000000000000000000000000000000604fbfc634eef00000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000038bf293f652b46fe85a15838d7ff736add1b6098ed1c143f3902869d325f9e0069e2b424053b9ebfcedf89ecb8bf2972974e98700c63af639e0000000000000000",
      "topics": [
        "0xa07a543ab8a018198e99ca0184c93fe9050a79400a0a723441f84de1d972cc17",
        "0x000000000000000000000000e2b424053b9ebfcedf89ecb8bf2972974e98700c",
      ],
    },
    {
      "address": "0x6b175474e89094c44da98b954eedeac495271d0f",
      "data": "0x00000000000000000000000000000000000000000000006c6b935b8bbd400000",
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x000000000000000000000000e2b424053b9ebfcedf89ecb8bf2972974e98700c",
        "0x0000000000000000000000009008d19f58aabd9ed0d60971565aa8510560ab41",
      ],
    },
  ],
  "txHash": "0xa7b2f2ff14a780504c97aeb484bfd7560486241c00f066395e1e8c2e9d99ffd3",
}
`;

exports[`Tenderly Simulator simulate() returns expected output on failed simulation 1`] = `undefined`;

exports[`Tenderly Simulator simulate() returns expected output on successful simulation 1`] = `undefined`;
